beluga_amcl/src/amcl_node_utils.cpp

```diff
@@ -50,4 +56,23 @@ std::vector<std::pair<double, double>> make_points_from_laser_scan(
   return points;
 }
 
+ROSOccupancyGrid make_occupancy_grid(beluga_amcl::messages::OccupancyGridConstSharedPtr ros_msg_ptr) {
+  auto make_origin_transform = [](const messages::Pose& origin) {
+    const auto rotation = Sophus::SO2d{tf2::getYaw(origin.orientation)};
+    const auto translation = Eigen::Vector2d{origin.position.x, origin.position.y};
+    return Sophus::SE2d{rotation, translation};
+  };
+
+  ROSOccupancyGrid::MapStorage grid_storage(ros_msg_ptr->info.width, ros_msg_ptr->info.height);
+  for (std::size_t x = 0; x < ros_msg_ptr->info.width; ++x) {
+    for (std::size_t y = 0; y < ros_msg_ptr->info.height; ++y) {
+      const auto index = x + y * ros_msg_ptr->info.width;
+      grid_storage.cell(static_cast<int>(x), static_cast<int>(y)) = ros_msg_ptr->data[index];
+    }
+  }
```

Comment:
@glpuga meta: one thing I don't love about this is that now we unconditionally copy the entire grid. That's OK if grids never update and performance improves. It's not that cool for large grids that change frequently or if there is no cache to leverage and memory is scarce (thinking of microcontrollers). Having a storage that maps an existing memory layout to our grids APIs would be nice. FWIW that's another reason why the grid hierarchy is mostly behavior with no state.

Reply:
I see the point, and nothing prevents us from creating a storage that just maps accesses to an existing buffer like we did before if needed, except lack of generality in the code above to leave initialization out of the grid storage scope to isolate it from ROS types.

I'm not too concerned about that kind of systems constrains (low memory + lack of cache) in the `beluga_amcl` code above too much, though; any system that would finds itself under those constraints is unlikely to run ROS in the first place.

Reply:
> any system that would finds itself under those constraints is unlikely to run ROS in the first place.

Definitely. As long as we don't propagate the constraint to the core library, I'm good :)